### PR TITLE
Encrypt potentially sensitive data

### PR DESCRIPTION
**Pull request by dcramer**
This adds some new column-based encryption facilities.

Using those it encrypts the following attributes:

- AuthIdentity.data
- AuthProvider.config
- Option.value
- OrganizationOption.value
- ProjectOption.value
- UserOption.value

### CODE REVIEW

**Comment by dcramer** [comment]
Ok so the primary issue today is that the implementation of encryption uses random noise to generate a new value each time. This means theres a couple choices to be made:

- Do we change encryption to require only mechanisms which dont have noise?
- Do we add an identifier field for each encrypted field (which needs indexed) to let us look it up? It'd be a hash of the unencrypted value.

Additionally we'd need to increase the length of some of these fields as the encrypted value is longer than the current allowed length.

**Comment by dstufft** [comment]
After thinking about this, I think my recommendation would be to add some sort of identifier. In an ideal world this would be a random, unique, value that has no relationship to the actual data, but given constraints I think it would be reasonable to use a hash for it.

The reason why I think the hash is the best answer is that while AES-CBC's (which is what Fernet uses) flaws when you reuse an IV is that you leak the fact that two encrypted pieces of data share the same prefix other kinds of constructs will have a much worse failure mode, up to, and including, being able to extract the complete key and decrypt everything.

In contrast, using a secure hash for identification doesn't leak anything about common prefixes, and the only thing it would actually leak is if two pieces of hashed data are *exactly* the same (or if they collide, which is possible to varying degrees of reality based on the size of the hash you select), but otherwise you can't get any information about how two hashed values relate to each other.

My suggestion for hashing would be to use blake2b (via [pyblake2](https://pypi.io/project/pyblake2/)) which is a very fast (faster than MD5) hash function with security properties similar to that of SHA3. It includes the concept of "hash personalization" and hash keying, which would let you ensure that not only do two different sentry installs not have the property that H(1234) == H(1234), but also that two different *columns* in Sentry do not have H(1234) == H(1234). So you would do something like:

```python
from pyblake2 import blake2b

# This value is secret and should vary per installation of Sentry
SENTRY_PER_INSTALL_KEY = b"some secret value that's unique to one sentry install"

def hash_apikey(apikey):
    # This value is not secret, it can be commited but it should vary per unique
    #  use of the hash function. For instance, if you hash both API Keys, and
    #  Usernames, then you want one value to be something like b"apikeys" and
    # the other value to be something like b"usernames". The value should not
    # be reused for different purposes in the same code base, you could even
    # prefix with b"sentry." for a bit of extra personalization.
    hash_personalization = b"sentry.apikey"
    h = blake2b(apikey, key=SENTRY_PER_INSTALL_KEY, person=hash_personalization)
    return h.hexdigest()
```

**Comment by dstufft** [comment]
Oh, and while blake2 allows you to select differently sized digests, I would recommend using the full size here.

File: src/sentry/db/models/fields/bounded.py [hunk 1]
```diff
@@ -110,3 +110,6 @@ class BoundedBigAutoField(BoundedAutoField):
 add_introspection_rules([], ["^sentry\.db\.models\.fields\.bounded\.BoundedBigIntegerField"])
 add_introspection_rules([], ["^sentry\.db\.models\.fields\.bounded\.BoundedPositiveIntegerField"])
 add_introspection_rules([], ["^sentry\.db\.models\.fields\.pickle\.UnicodePickledObjectField"])
```
Comment (dcramer) [inline src/sentry/db/models/fields/bounded.py]: of note these should not be in this file (just realized where it was)

File: src/sentry/models/projectkey.py [hunk 2]
```diff
@@ -36,8 +36,8 @@ class ProjectKeyStatus(object):
 class ProjectKey(Model):
     project = FlexibleForeignKey('sentry.Project', related_name='key_set')
     label = models.CharField(max_length=64, blank=True, null=True)
-    public_key = models.CharField(max_length=32, unique=True, null=True)
-    secret_key = models.CharField(max_length=32, unique=True, null=True)
+    public_key = EncryptedCharField(max_length=32, unique=True, null=True)
```
Comment (dcramer) [inline src/sentry/models/projectkey.py]: needs much larger bounds on length

File: src/sentry/models/projectkey.py [hunk 3]
```diff
@@ -36,8 +36,8 @@ class ProjectKeyStatus(object):
...
+    secret_key = EncryptedCharField(max_length=32, unique=True, null=True)
```
Comment (dcramer) [inline src/sentry/models/projectkey.py]: needs much larger bounds on length

File: src/sentry/models/apitoken.py [hunk 4]
```diff
@@ -8,15 +8,15 @@
...
+    Model, BaseManager, EncryptedCharField, FlexibleForeignKey, sane_repr
 )
 
 
 class ApiToken(Model):
     # users can generate tokens without being key-bound
     key = FlexibleForeignKey('sentry.ApiKey', null=True)
     user = FlexibleForeignKey('sentry.User')
-    token = models.CharField(max_length=64, unique=True)
+    token = EncryptedCharField(max_length=64, unique=True)
```
Comment (dcramer) [inline src/sentry/models/apitoken.py]: needs much larger bounds on length

File: src/sentry/models/apikey.py [hunk 5]
```diff
@@ -30,7 +30,7 @@ class ApiKeyStatus(object):
 class ApiKey(Model):
     organization = FlexibleForeignKey('sentry.Organization', related_name='key_set')
     label = models.CharField(max_length=64, blank=True, default='Default')
-    key = models.CharField(max_length=32, unique=True)
+    key = EncryptedCharField(max_length=32, unique=True)
```
Comment (dcramer) [inline src/sentry/models/apikey.py]: needs much larger bounds on length

**Comment by dcramer** [comment]
Going to shift towards hash lookups:

- Add a column which gets hashed from the encrypted value column (can automatically bind the value)
- When querying for one of these columns, utilize the other named field
- If a named field is not specified, the column cannot be queried against using the ORM

Something like:

```
foo = EncryptedCharField(hash_field='foo_hash')
foo_hash = HashField(data_field='foo')
```

File: src/sentry/models/projectkey.py (hunk 3 above)
Comment (mattrobenolt) [inline src/sentry/models/projectkey.py]: I'd probably say these Encrypted fields should just be `TextField` based unless the length can reliably be determined.

Either way, to Postgres at least, a `char` type is the same as `text` just with a constraint, so I think we can live without a length constraint here imo.

We'd also likely need to move this `unique` indexes to the new hashed fields since I think we determined that even identical values should be encrypted into different values, so uniqueness could only apply to the hash.

Comment (dstufft) [inline src/sentry/models/projectkey.py]: For a block cipher like AES the size will be whatever the size of the input is rounded up to 8 bytes (or if it's already a multiple of 8 bytes, with an extra 8 bytes on the end). For fernet there's also additional overhead for things like the version tag, the authentication tag, the IV, etc. 

For a stream cipher, the output will be equal to the input in size (plus any of the additional overhead like above).

Given all of that, I agree with @mattrobenolt you should just use a ``text`` type and a ``unique`` index on the hashed fields.

**Comment by graingert** [comment]
ApiTokens can be stored as psuedo username: hashed password pairs.

just extend AbstractBaseUser, when generating an api key use:
```python
def reset_api_key(self):
    api_key_key = get_random_string(8)
    api_key_secret = get_random_string(35)
    api_key.key = api_key_key
    api_key.set_password(api_key_secret)

return f"{api_key_key}-{api_key_secret}"
```

To check a key, you just do:
```python
def check_api_key(key):
    key, password = key.partition('-')
    ApiKey.get(key=key).check_password(password)
```

To migrate your old keys you just need a data migration:

```python
def migrate_api_key(api_key: ApiKey) -> None:
    key, password = api_key.key.partition('-')
    api_key.key = key
    api_key.set_password(password)
```

This also enforces that api tokens are only visible once

**Comment by dcramer** [comment]
I think we'll get some of the basics of this in so we can at least support encrypted values in the DB, but we likely wont migrate all fields at this time.

**Comment by graingert** [comment]
But most of your data should be hashed not encrypted

**Comment by dcramer** [comment]
@graingert We store API keys, passwords, etc and those cannot ever be hashed if they're used to authenticate with another service.

**Comment by graingert** [comment]
Sure, but last time I checked this code it was encrypting API keys generated by Sentry. Those should be hashed instead

**Comment by dcramer** [comment]
@graingert we can agree to disagree. It's a poor user experience in many situations to do that, and often the keys are not sensitive enough to warrant it.

**Comment by graingert** [comment]
I'm not sure why it would be a good user experience to reuse an API key after generation. It would be better to offer an easy way to regenerate them.

Some examples of services that are unable to show API keys after generation: AWS IAM, Google App passwords

**Comment by dcramer** [comment]
@graingert those same examples also show API keys after generation in many cases. Either way, it's not a debate, we're not going to hash DSNs and never let you access the value after generation. I'd like to keep this kind of noise out of the PR as its not very relevant, and we're already extremely security conscious here at Sentry.

**Comment by graingert** [comment]
Sure, can you have a look at my PRs?

**Comment by dcramer** [comment]
This is ready to go. We can followup with more features later.

File: src/sentry/models/groupsubscription.py [hunk 6]
```diff
@@ -88,19 +88,21 @@ def get_participants(self, group):
 
         # Find users which by default do not subscribe.
         participating_only = set(
-            UserOption.objects.filter(
+            uo.user_id for uo in UserOption.objects.filter(
```
Comment (dcramer) [inline src/sentry/models/groupsubscription.py]: @tkaemming kind of annoying, but i think still worth it for sane defaults on the model

File: src/sentry/db/models/fields/encrypted.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    def get_prep_lookup(self, lookup_type, value):
+        raise NotImplementedError('{!r} lookup type for {!r} is not supported'.format(
+            lookup_type,
+            self,
+        ))
+
+
+class EncryptedPickledObjectField(PickledObjectField):
+    def get_db_prep_value(self, value, *args, **kwargs):
+        if isinstance(value, six.binary_type):
```
Comment (mattrobenolt) [inline src/sentry/db/models/fields/encrypted.py]: This seems odd to me that we're coercing all binary values into utf-8 strings.

Same on the `to_python` side, we're explicitly want `string_type` spit out instead of raw binary. I feel like I'd expect the exact opposite. Raw bytes in, raw bytes out. Up to the caller to handle encoding/decoding.

Then these fields can subclass `BinaryField` iirc.

Comment (mitsuhiko) [inline src/sentry/db/models/fields/encrypted.py]: Yeah, why do we need to do this at all. I think the pickle field supports them both separately anyways.

**Comment by mattrobenolt** [comment]
Do we have _any_ idea how much CPU overhead we're adding for these fields? Some of them are really high touch values that are handled a lot. Specifically `ProjectOption.value`, `OrganizationOption.value` and `Option.value`. Moreso, `ProjectOption` and `OrganizationOption` concern me because of how we bulk fetch all of the options whether we need them or not. I think this will be on the order of dozens or into 20's of rows on each `/store/` call. So would be good to have a ballpark idea of the impact this will have now decrypting all of these values and adding O(n) work to something that used to be considered free.

Alternatively, would be nice to expose the decrypted values as lazy values so they're not decrypted when pulled from the database, but decrypted when the value is actually accessed.

File: src/sentry/utils/encryption.py [hunk 8]
```diff
@@ -28,18 +28,22 @@ def encrypt(self, value):
...
-            pass
-        else:
-            enc_data = smart_bytes(enc_data)
-            try:
-                scheme = self.schemes[enc_method]
-            except KeyError:
-                raise ValueError('Unknown encryption scheme: %s'.format(enc_method))
-            value = scheme.decrypt(enc_data)
-        return value
+            return value
```
Comment (mitsuhiko) [inline src/sentry/utils/encryption.py]: Is this done to attempt to support old values? Do we need to do this?

File: src/sentry/db/models/fields/encrypted.py (hunk 7 above)
Comment (dcramer) [inline src/sentry/db/models/fields/encrypted.py]:
```
class UnicodePickledObjectField(PickledObjectField):
    def get_db_prep_value(self, value, *args, **kwargs):
        if isinstance(value, six.binary_type):
            value = value.decode('utf-8')
        return super(UnicodePickledObjectField, self).get_db_prep_value(
            value, *args, **kwargs)
```

File: src/sentry/utils/encryption.py (hunk 8 above)
Comment (dcramer) [inline src/sentry/utils/encryption.py]: @mitsuhiko its to support transitional values, or values that look encrypted but arent actually (due to have a $ in them).

Comment (mitsuhiko) [inline src/sentry/utils/encryption.py]: I'm not super opposed to doing that but it seems weird that we support (some) unencrypted data in these fields.

Comment (dcramer) [inline src/sentry/utils/encryption.py]: the only way to implement this is to support unencrypted values -- otherwise all existing values will be broken

Comment (mitsuhiko) [inline src/sentry/utils/encryption.py]: Hmm. I thought you planned on making a data migration and encrypting them all.

Comment (dcramer) [inline src/sentry/utils/encryption.py]: Even if that were true it wouldnt be synchronous so we'd have to have an intermittent phase.

tl;dr you should be able to enable encryption without downtime, and then backfill encrypted values

Comment (mitsuhiko) [inline src/sentry/utils/encryption.py]: I agree with that, i just feel like this detection will be insufficient for actually detecting encrypted values reliably but I might be be too pessimistic here.

**Review by mattrobenolt — APPROVED** [review]
wgcw

**Comment by dcramer** [comment]
@mitsuhiko i made tweaks to core encryption (since its unused) to add the prefix. let me know if you have any other concerns before landing

**Review by mitsuhiko — APPROVED** [review]
lgtm

File: src/sentry/db/models/fields/bounded.py (hunk 1 above)
Comment (mattrobenolt) [inline src/sentry/db/models/fields/bounded.py]: Were you going to remove this? They seem to have been added into the `fields/encrypted.py`.

Comment (dcramer) [inline src/sentry/db/models/fields/bounded.py]: so i think this i no longer valid as i couldnt find it in local code, but i did fix the missing ones

Comment (dcramer) [inline src/sentry/db/models/fields/bounded.py]: NEVERMIND

**Comment by dcramer** [comment]
changed marker/fixed binary issues

**Review by mitsuhiko — APPROVED** [review]
This seems reasonable.
